Improve io registration

fbudin69500 opened the pull request.

fbudin69500 committed:
ENH: Improving error message in ITKImageFileWriter

When catching an ITK exception in ITKImageFileWriter,
the exception description is now printed in the error message
displayed in DREAMD3D.

fbudin69500 committed:
BUG: Verifies data array path in ITKImageWriter

Verifies that the given dataArrayPath actually exists. Using the UI
to select the dataArrayPath to save does not allow to select a
path that does not exist. However, the default path at which the
variable is set may not exists. If one runs the filter without
changing that default path, no check was perform and DREAM3D was crashing.

fbudin69500 committed:
ENH: Registering IO only once, when loading plugin

Registering the IO in entry point of the plugin. This allows
to register the IO only once. Tests have been improved to
support 2D images (to test SCIFIO formats like TIF or PNG).

fbudin69500 committed:
ENH: Improve Reader and Writer filters UI

Default values consistent with DREAM3D's default values.

fbudin69500 committed:
BUG: Missing Graft() overloaded function

Graft() function had not been reimplemented in itk::Dream3DImage.
When Graft() was called, it is the implementation available in
itk::Image that was called, which lead to filters crashing due to the
fact that it was casting the PixelContainerType to be casted in
the wrong type: ImportImageContainer instead of
ImportDream3DImageContainer.

fbudin69500 committed:
ENH: Registering TIFF IO and adding tests for PNGIO

Registration of the TIFF library to read and write TIFF images
with ITK from DREAM3D.
Addition of tests for ITKImageReader to test loading of PNG and TIFF
images.

thewtex commented on ITKImageProcessingFilters/itkDream3DImage.hxx:
```diff
@@ -102,6 +102,34 @@ namespace itk
+        this->SetPixelContainer(const_cast< PixelContainerType * >
+          (imgData->GetPixelContainer()));
+      }
+      else
```
Please use the itkDynamicCastInDebugMode macro here to avoid the template dynamic_cast issues.

fbudin69500 replied:
Does this mean the code in [ITKImage.hxx](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImage.hxx#L116) should also be modified?

thewtex commented on ITKImageProcessingPlugin.cpp:
```diff
@@ -59,7 +73,18 @@ m_Copyright("Copyright"),                               // Initialize ITKImagePr
+  itk::BioRadImageIOFactory::RegisterOneFactory();
+  itk::GE4ImageIOFactory::RegisterOneFactory();
+  itk::GE5ImageIOFactory::RegisterOneFactory();
+  itk::SCIFIOImageIOFactory::RegisterOneFactory();
```
8-)

thewtex commented:
Nice!

One inline comment.

thewtex commented on ITKImageProcessingFilters/itkDream3DImage.hxx:
```diff
@@ -102,6 +102,34 @@ namespace itk
+        this->SetPixelContainer(const_cast< PixelContainerType * >
+          (imgData->GetPixelContainer()));
+      }
+      else
```
On Sat, Sep 3, 2016 at 1:45 PM, Francois Budin notifications@github.com
wrote:

> In ITKImageProcessingFilters/itkDream3DImage.hxx
> https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/20#discussion_r77440400
> :
> 
> > -  {
> > -    // call the superclass' implementation
> > -    Superclass::Superclass::Graft(data);
> >   +
> > -    if (data)
> > -    {
> > -      // Attempt to cast data to an Image
> > -      const Self \* const imgData = dynamic_cast< const Self \* >(data);
> >   +
> > -      if (imgData != ITK_NULLPTR)
> > -      {
> > -        // Now copy anything remaining that is needed
> > -        this->SetPixelContainer(const_cast< PixelContainerType \* >
> > -          (imgData->GetPixelContainer()));
> > -      }
> > -      else
> 
> Does this mean the code in ITKImage.hxx
> https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImage.hxx#L116
> should also be modified?
> 
> Yes

fbudin69500 committed:
BUG: Replace dynamic_cast with static_cast

Replacing dynamic_cast with static_cast (itkDynamicCastInDebugMode) since
static_cast should always be valid in this section. This allows to prevent
some errors on some compilers which do not support the use of dynamic_cast
with templated classes instantiated in a translation unit different than
the one where they are defined [1].

[1] https://github.com/Slicer/Slicer/pull/573#issuecomment-244484002

fbudin69500 commented:
Closing this pull-request as these commits were also part of https://github.com/BlueQuartzSoftware/ITKImageProcessing/pull/21
